### ISSUE TITLE
nan is not a known member - pyright 1.1.340 with numpy stubs

### ISSUE THREAD

**Issue by fheinze-tkb** [microsoft/pyright#6727]
<h3> Environment data </h3>

-   Language Server version: 2023.11.104
-   Pyright version 1.1.340
-   OS and version: linux x64 / WSL in Windows 10
-   Python version (and distribution if applicable, e.g. Anaconda):  3.12, pyenv
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

**pyproject.toml**

```
[tool.poetry]
name = "numpy-issue"
version = "0.1.0"
description = ""


[tool.poetry.dependencies]
python = "^3.12"
numpy = "^1.26.2"


[tool.poetry.group.dev.dependencies]
pyright = "^1.1.340"

[tool.pyright]
stubPath="./stubs"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```

<h3> Code Snippet </h3>

```python
import numpy as np

print(np.nan)
```

<h3> Repro Steps </h3>

1. Setup a new project with poetry
2. Add numpy and pyright as dependencies
3. Copy stubs for numpy from pylance-extension into project folder and link it in pyproject.toml
4. run pyright for code snippet-file

<h3> Expected behavior </h3>
0 errors, 0 warnings, 0 informations

This worked fine for our latest pyright version: 1.1.333 and it still works fine with pylance.

<h3> Actual behavior </h3>
It will throw the following error:

```
error: "nan" is not a known member of module "numpy" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```
We integrated stubs because of the following issue: https://github.com/microsoft/pylance-release/issues/5005

<h3> Logs </h3>

```
2023-12-13 15:21:43.689 [info] (Client) Pylance async client (2023.11.104) started with python extension (2023.22.0)
2023-12-13 15:21:44.470 [info] [Info  - 3:21:44 PM] (110516) Pylance language server 2023.11.104 (pyright a78e2b10) starting
2023-12-13 15:21:44.471 [info] [Info  - 3:21:44 PM] (110516) Server root directory: /home/fheinze/.vscode-server/extensions/ms-python.vscode-pylance-2023.11.104/dist
2023-12-13 15:21:44.487 [info] [Info  - 3:21:44 PM] (110516) Starting service instance "numpy-issue"
2023-12-13 15:21:44.581 [info] (110516) No configuration file found.
2023-12-13 15:21:44.582 [info] (110516) pyproject.toml file found at /home/fheinze/Code/playground/py/numpy-issue.
2023-12-13 15:21:44.584 [info] [Info  - 3:21:44 PM] (110516) Setting pythonPath for service "numpy-issue": "/home/fheinze/Code/playground/py/numpy-issue/.venv/bin/python"
2023-12-13 15:21:44.584 [info] [Info  - 3:21:44 PM] (110516) Setting environmentName for service "numpy-issue": "3.12.0 (.venv venv)"
2023-12-13 15:21:44.597 [info] [Info  - 3:21:44 PM] (110516) Loading pyproject.toml file at /home/fheinze/Code/playground/py/numpy-issue/pyproject.toml
2023-12-13 15:21:44.645 [info] [Info  - 3:21:44 PM] (110516) Assuming Python version 3.12
2023-12-13 15:21:44.646 [info] (110516) Assuming Python platform Linux
2023-12-13 15:21:44.727 [info] [Info  - 3:21:44 PM] (110516) Search paths for /home/fheinze/Code/playground/py/numpy-issue
2023-12-13 15:21:44.728 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/.vscode-server/extensions/ms-python.vscode-pylance-2023.11.104/dist/typeshed-fallback/stdlib
2023-12-13 15:21:44.728 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/Code/playground/py/numpy-issue
2023-12-13 15:21:44.728 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/Code/playground/py/numpy-issue/src
2023-12-13 15:21:44.729 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/Code/playground/py/numpy-issue/stubs
2023-12-13 15:21:44.729 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/.vscode-server/extensions/ms-python.vscode-pylance-2023.11.104/dist/typeshed-fallback/stubs/...
2023-12-13 15:21:44.729 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/.vscode-server/extensions/ms-python.vscode-pylance-2023.11.104/dist/bundled/stubs
2023-12-13 15:21:44.730 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/.pyenv/versions/3.12.0/lib/python3.12
2023-12-13 15:21:44.731 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/.pyenv/versions/3.12.0/lib/python3.12/lib-dynload
2023-12-13 15:21:44.731 [info] [Info  - 3:21:44 PM] (110516)   /home/fheinze/Code/playground/py/numpy-issue/.venv/lib/python3.12/site-packages
2023-12-13 15:21:44.731 [info] [Info  - 3:21:44 PM] (110516) Adding fs watcher for library directories:
 /home/fheinze/.pyenv/versions/3.12.0/lib/python3.12
/home/fheinze/.pyenv/versions/3.12.0/lib/python3.12/lib-dynload
/home/fheinze/Code/playground/py/numpy-issue/.venv/lib/python3.12/site-packages
2023-12-13 15:21:44.732 [info] [Info  - 3:21:44 PM] (110516) Adding fs watcher for directories:
 /home/fheinze/Code/playground/py/numpy-issue
2023-12-13 15:21:44.732 [info] (110516) Searching for source files
2023-12-13 15:21:44.743 [info] [Info  - 3:21:44 PM] (110516) Found 16 source files
2023-12-13 15:21:44.743 [info] (110516) pytest configurations: {"message":"script","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2023-12-13 15:21:44.744 [info] (110516) pytest configurations: {"message":"script","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2023-12-13 15:21:45.374 [info] (110516) pytest configurations: {"message":"request cancelled","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2023-12-13 15:21:45.454 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/src/main.py ...
2023-12-13 15:21:45.455 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/src/main.py [fs read 0ms] (76ms)
2023-12-13 15:21:45.669 [info] (110516) [BG(1)]   parsing: /home/fheinze/.vscode-server/extensions/ms-python.vscode-pylance-2023.11.104/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (212ms)
2023-12-13 15:21:45.747 [info] (110516) [BG(1)]   binding: /home/fheinze/.vscode-server/extensions/ms-python.vscode-pylance-2023.11.104/dist/typeshed-fallback/stdlib/builtins.pyi (77ms)
2023-12-13 15:21:45.747 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/src/main.py (1ms)
2023-12-13 15:21:45.749 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/src/main.py [found 0] (370ms)
2023-12-13 15:21:45.870 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_multiarray_umath.pyi ...
2023-12-13 15:21:45.870 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_multiarray_umath.pyi [fs read 5ms] (120ms)
2023-12-13 15:21:45.943 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_multiarray_umath.pyi (73ms)
2023-12-13 15:21:45.982 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_multiarray_umath.pyi [found 220] (233ms)
2023-12-13 15:21:45.996 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_simd.pyi ...
2023-12-13 15:21:45.997 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_simd.pyi [fs read 1ms] (1ms)
2023-12-13 15:21:45.997 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_simd.pyi (0ms)
2023-12-13 15:21:45.997 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/core/_simd.pyi [found 2] (2ms)
2023-12-13 15:21:45.997 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/fft/_pocketfft_internal.pyi ...
2023-12-13 15:21:45.998 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/fft/_pocketfft_internal.pyi [fs read 1ms] (1ms)
2023-12-13 15:21:45.998 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/fft/_pocketfft_internal.pyi (0ms)
2023-12-13 15:21:45.998 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/fft/_pocketfft_internal.pyi [found 2] (2ms)
2023-12-13 15:21:45.998 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/_umath_linalg.pyi ...
2023-12-13 15:21:45.998 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/_umath_linalg.pyi [fs read 1ms] (3ms)
2023-12-13 15:21:45.998 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/_umath_linalg.pyi (2ms)
2023-12-13 15:21:46.001 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/_umath_linalg.pyi [found 22] (9ms)
2023-12-13 15:21:46.005 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/lapack_lite.pyi ...
2023-12-13 15:21:46.006 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/lapack_lite.pyi [fs read 0ms] (3ms)
2023-12-13 15:21:46.006 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/lapack_lite.pyi (1ms)
2023-12-13 15:21:46.007 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/linalg/lapack_lite.pyi [found 10] (4ms)
2023-12-13 15:21:46.008 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_bounded_integers.pyi ...
2023-12-13 15:21:46.009 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_bounded_integers.pyi [fs read 0ms] (1ms)
2023-12-13 15:21:46.009 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_bounded_integers.pyi (0ms)
2023-12-13 15:21:46.009 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_bounded_integers.pyi [found 4] (2ms)
2023-12-13 15:21:46.013 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_common.pyi ...
2023-12-13 15:21:46.013 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_common.pyi [fs read 0ms] (4ms)
2023-12-13 15:21:46.015 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_common.pyi (1ms)
2023-12-13 15:21:46.015 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_common.pyi [found 6] (6ms)
2023-12-13 15:21:46.038 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_generator.pyi ...
2023-12-13 15:21:46.039 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_generator.pyi [fs read 1ms] (23ms)
2023-12-13 15:21:46.048 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_generator.pyi (9ms)
2023-12-13 15:21:46.052 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_generator.pyi [found 8] (36ms)
2023-12-13 15:21:46.055 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_mt19937.pyi ...
2023-12-13 15:21:46.056 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_mt19937.pyi [fs read 0ms] (3ms)
2023-12-13 15:21:46.056 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_mt19937.pyi (1ms)
2023-12-13 15:21:46.057 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_mt19937.pyi [found 4] (5ms)
2023-12-13 15:21:46.060 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_pcg64.pyi ...
2023-12-13 15:21:46.060 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_pcg64.pyi [fs read 0ms] (3ms)
2023-12-13 15:21:46.061 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_pcg64.pyi (0ms)
2023-12-13 15:21:46.061 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_pcg64.pyi [found 4] (4ms)
2023-12-13 15:21:46.064 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_philox.pyi ...
2023-12-13 15:21:46.065 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_philox.pyi [fs read 0ms] (3ms)
2023-12-13 15:21:46.065 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_philox.pyi (1ms)
2023-12-13 15:21:46.067 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_philox.pyi [found 5] (5ms)
2023-12-13 15:21:46.069 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_sfc64.pyi ...
2023-12-13 15:21:46.069 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_sfc64.pyi [fs read 1ms] (3ms)
2023-12-13 15:21:46.070 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_sfc64.pyi (0ms)
2023-12-13 15:21:46.070 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/_sfc64.pyi [found 4] (4ms)
2023-12-13 15:21:46.085 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/bit_generator.pyi ...
2023-12-13 15:21:46.086 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/bit_generator.pyi [fs read 0ms] (15ms)
2023-12-13 15:21:46.088 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/bit_generator.pyi (2ms)
2023-12-13 15:21:46.092 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/bit_generator.pyi [found 17] (20ms)
2023-12-13 15:21:46.121 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/mtrand.pyi ...
2023-12-13 15:21:46.121 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/mtrand.pyi [fs read 1ms] (30ms)
2023-12-13 15:21:46.133 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/mtrand.pyi (11ms)
2023-12-13 15:21:46.144 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/stubs/numpy/random/mtrand.pyi [found 57] (52ms)
2023-12-13 15:21:46.147 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/tests/__init__.py ...
2023-12-13 15:21:46.147 [info] (110516) [BG(1)]   parsing: /home/fheinze/Code/playground/py/numpy-issue/tests/__init__.py [fs read 1ms] (2ms)
2023-12-13 15:21:46.147 [info] (110516) [BG(1)]   binding: /home/fheinze/Code/playground/py/numpy-issue/tests/__init__.py (0ms)
2023-12-13 15:21:46.147 [info] (110516) [BG(1)] indexing: /home/fheinze/Code/playground/py/numpy-issue/tests/__init__.py [found 0] (2ms)
2023-12-13 15:21:46.163 [info] (110516) pytest configurations: {"message":"script","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
```

**Comment by debonte** [comment]
That empty URI is explicitly pushed in there in the following if block:

```typescript
                    if (!isNativeLib && foundDirectory) {
                        importFailureInfo.push(`Partially resolved import with directory '${dirPath}'`);
                        resolvedPaths.push(Uri.empty());
                        if (isLastPart) {
                            implicitImports = this._findImplicitImports(importName, dirPath, [pyFilePath, pyiFilePath]);
                            isNamespacePackage = true;
                        }
                    }
```

**Comment by rchiodo** [comment]
I would compare against what we had before the URI change. Maybe I missed a check to see if something was empty or not. That resolvedPaths.push looks to me like it should push the dirPath, but I'd double check what was there before.

**Comment by rchiodo** [comment]
It seems it used to push empty there, so I guess that's correct:
https://github.com/microsoft/pyrx/blob/4c83a345897ca58625c17a338c73a158984aeb08/packages/pyright/packages/pyright-internal/src/analyzer/importResolver.ts#L1453

**Comment by debonte** [comment]
Yeah, I guess that's intentional as it's "partially resolved". I'll keep comparing...

**Comment by rchiodo** [comment]
Else clause was the same before too. Generally when I was fixing these sort of things before I submitted, I had to run both versions side by side to see where they diverged.

**Comment by erictraut** [comment]
@rchido & @debonte, is this fixed?

**Comment by debonte** [comment]
I don't believe so. I'm planning to look at this more tomorrow.

**Comment by erictraut** [comment]
OK, thanks. Once you have this fixed, I may publish a new release of pyright to address all of the URI-related regressions that we've found and fixed since 1.1.340 was released on Tuesday.

**Comment by debonte** [comment]
Thanks for the bug report. This will be addressed in the next release.

This was caused by the change to add URI support.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.341, which I just published.
